subsys/net/lib/app/net_app.c

```diff
@@ -1247,6 +1247,21 @@ int _net_app_tls_sendto(struct net_pkt *pkt,
 	struct k_mem_block block;
 	int ret;
 
+	if (!ctx->tls.handshake_done) {
+		/* This means that the initial TLS handshake is not yet
```

Comment:
IMHO, it would be better to make TLS handshake part of the connection process (i.e. connection is done when the handshake is finished, and only then user code can run). I understand it may be not immediately easy/possible with the current code.

Reply:
The code actually does it like that, connection flow continues after TLS handshake if finished.
The send function can be called by application before the TLS thread has managed to establish the connection. This check is for that scenario.